2020/4/passport/src/main.rs

```diff
@@ -0,0 +1,248 @@
+                }
+                "cid" => {
+                    country_id = Some(String::from(value));
+                }
+
+                &_ => {}
+            }
+        }
+        Passport {
+            birth_year: birth_year,
+            issue_year: issue_year,
+            expiration_year: expiration_year,
+            height: height,
+            hair_color: hair_color,
+            eye_color: eye_color,
+            passport_id: passport_id,
+            country_id: country_id,
+        }
+    }
+
+    pub fn is_valid(&self) -> bool {
+        match &self.birth_year {
+            Some(byr) => is_valid_year(&byr, 1920, 2002) && match &self.issue_year {
+                Some(iyr) => is_valid_year(&iyr, 2010, 2020) && match &self.expiration_year {
+                    Some(eyr) => is_valid_year(&eyr, 2020, 2030) && match &self.height {
+                        Some(h) => is_valid_height(h) && match &self.hair_color {
+                            Some(hcl) => is_valid_hair_color(hcl) && match &self.eye_color {
+                                Some(ecl) => is_valid_eye_color(ecl) && match &self.passport_id {
+                                    Some(pid) => is_valid_passport_id(pid),
+                                    None => false,
+                                },
+                                None => false,
+                            },
+                            None => false,
+                        },
+                        None => false,
+                    },
+                    None => false,
+                },
+                None => false,
+            },
+            None => false,
+        }
+    }
+}
+
+fn is_valid_year(value: &str, min: usize, max: usize) -> bool {
+    if value.len() != 4 {
+        return false;
+    }
+
+    let v = value.parse::<usize>().unwrap();
+    min <= v && v <= max
+}
+
+fn is_valid_hair_color(value: &str) -> bool {
+    if value.len() != 7 {
+        return false;
+    }
+
+    let mut chars = value.chars();
+    if chars.next().unwrap() != '#' {
+        return false;
+    }
+
+    for _ in 1..6 {
+        if !chars.next().unwrap().is_digit(16) {
+            return false;
+        }
+    }
+    true
+}
+
+fn is_valid_eye_color(value: &str) -> bool {
+    match value {
+        "amb" => true,
+        "blu" => true,
+        "brn" => true,
+        "gry" => true,
+        "grn" => true,
+        "hzl" => true,
+        "oth" => true,
+        _ => false,
+    }
+}
+
+fn is_valid_passport_id(value: &str) -> bool {
+    value.len() == 9 && value.chars().all(|c| c.is_digit(10))
+}
+
+fn is_valid_height(value: &str) -> bool {
+    if value.ends_with(&"cm") {
+        return is_between(value, 150, 193);
+    } else if value.ends_with(&"in") {
+        return is_between(value, 59, 76);
+    }
+
+    false
+}
+
+fn is_between(value: &str, min: usize, max: usize) -> bool {
+    let len = value.len();
+    let v = value[0..len - 2].parse::<usize>();
+    match v {
```

Comment:
I might use an `if let` here instead of a full blown match:

```rust
if let Ok(vv) = v {
    vv >= min && vv <= max
} else {
    false
}
```

When there are only two arms, it feels a little simpler to me. Not sure if that's more "idiomatic" but I prefer it.